Column naming

luis901101 opened the pull request:
I made this PR based on [issue 67](https://github.com/conduit-dart/conduit/issues/67).

### Current situation
- Column doesn't allow to specify a custom name.
- Column names are not using proper naming conventions. 
    - e.g: Model field: `creationDate` => resulting column name: `creationdate`
    - e.g: Model field: `streetAddressLocation` => resulting column name: `streetaddresslocation`
- Table names are not using proper naming conventions. 
    - e.g: Model: `_User` => resulting table name: `_user`
    - e.g: Model: `_AccessPoint` => resulting table name: `_accesspoint`
- API response's JSON field names uses the Model field names; devs should be able to specify custom field name for JSON input/output to ensure proper naming convention. Model field names and JSON field names not necessarily should use the same naming convention. Although this can be done by overriding `readFromMap(...)` and `asMap()` from `ManagedObject` it will be a nightmare for devs to achieve this.
- API response's JSON always includes all fields no matter if it's null. Sometimes is necessary to avoid including null values in responses to reduce payload.

### What this PR brings
**NOTE:** All Column and Table naming conventions changes I mention here are considered **breaking changes**, anyway I managed to keep code legacy compatible with backguard compatibility, **using conduit would be the same**, nothing changes by default.

- `@Column()` updated to support custom naming and proper naming conventions with legacy support by default. By default `@Column()` remains the same, but now you can use `@Column(name: 'some_column_name')` to specify a custom name for this model field in the table. On the other hand you can use `@Column(useSnakeCaseName: true|false)`, to indicate to use **snake_case** naming, when this is false legacy naming will be used for naming the column. By default property `useSnakeCaseName` is `null` which delegates naming to `@Table()` *explained below*.
    - e.g: 
    ```dart
    @Column(nullable: true, name: 'creation_date')
    DateTime? creationDate;
    ```
    *Resulting column name: `creation_date`*
    - e.g: 
    ```dart
    @Column(nullable: true, useSnakeCaseName: true)
    String? streetAddressLocation;
    ```
    *Resulting column name: `street_address_location`*
    - e.g: 
    ```dart
    @Column(nullable: true)
    DateTime? creationDate;
    ```
    *Resulting column name: `creationdate`. This is the current situation and default*

- `@Table()` updated to support proper naming conventions with legacy support by default. By default `@Table()` remains the same, but now you can use `@Table(useSnakeCaseName: true|false)` to indicate to use **snake_case** naming, when this is false legacy naming will be used for naming the table; `useSnakeCaseName` is `false` by default. Also you can use `@Table(useSnakeCaseColumnName: true|false)` to indicate to use **snake_case** naming, when this is false legacy naming will be used for naming all the columns in the specified table; `useSnakeCaseColumnName` is `false` by default and can be overriden by `@Column(useSnakeCaseName: true|false)`.    
    - e.g: 
    ```dart
    @Table(useSnakeCaseName: true, useSnakeCaseColumnName: true)
    class _AccessPoint {
        @Column(nullable: true)
        String? mainStreet;
    }
    ```
    *Resulting table name: `access_point` and column name: `main_street`*
    - e.g: 
    ```dart
    @Table(useSnakeCaseName: true, useSnakeCaseColumnName: true)
    class _AccessPoint {
        @Column(nullable: true, useSnakeCaseName: false)
        String? mainStreet;
    }
    ```
    *Resulting table name: `access_point` and column name: `mainstreet`. Here `@Column(...)` overrides column legacy naming*
    - e.g: 
    ```dart
    @Table()
    class _AccessPoint {
        @Column(nullable: true)
        String? mainStreet;
    }
    ```
    *Resulting table name: `_accesspoint` and column name: `mainstreet`. This is the current situation and default*

- Created new annotation `@ResponseKey(...)` to allow devs to use custom API response's JSON field names and to decide wether or not a null field should be included in response. `@ResponseKey(...)` can be used with `@Column()` and `@Serialize()` for transient properties.
    - e.g: 
    ```dart
    @ResponseKey(name: 'CrEaTiOn_DaTe', includeIfNull: false)
    @Column(nullable: true)
    DateTime? creationDate;
    ```
    *Column name `creationdate`, resulting JSON field name: `"CrEaTiOn_DaTe": "2021-11-10T00:02:37.472299Z"`, if there is no data available for this field then will not be included in response*
    - e.g: 
    ```dart    
    @ResponseKey(includeIfNull: true)
    @Column(nullable: true, name: 'creation_date')
    DateTime? creationDate;
    ```
    *Column name `creation_date`, resulting JSON field name: `"creation_date": "2021-11-10T00:02:37.472299Z"`, if there is no data available for this field `null` value will be included in response. Here notes how JSON field name will match Column name*
    - e.g: 
    ```dart    
    @Column(nullable: true)
    DateTime? creationDate;
    ```
    *Column name `creationdate`, resulting JSON field name: `"creationDate": "2021-11-10T00:02:37.472299Z"`. Here notes how JSON field name will not match Column name, instead uses Model field name. this is the current situation and default*
    - e.g: 
    ```dart
    @Serialize()
    @ResponseKey(name: 'extra_info')
    String? get extraInfo => '$info and Some extra info';

    @Serialize()
    @ResponseKey(name: 'extra_info')
    set extraInfo(String? extra) => info = '$info $extra';
    ```
    *Resulting JSON field name: `"extra_info": "info from db and Some extra info"`*

- Created new annotation `@ResponseModel()` to allow devs to decide wether or not all null fields from a Model should be included in response. By default `@ResponseModel()` includes null fields in responses, to avoid including null fields in responses just `@ResponseModel(includeIfNullField: false)`. If a Model field uses anotation `@ResponseKey(includeIfNull: true|false)` then `@ResponseModel(...)` is overriden.

j4qfrost commented:
Thanks, I'll take a look this weekend. Currently integrating melos on Reductions suggestion.

Reductions commented on conduit/lib/src/db/managed/object.dart:
```diff
@@ -67,6 +66,42 @@ abstract class ManagedBacking {
 ///
 /// See more documentation on defining a data model at http://conduit.io/docs/db/modeling_data/
 abstract class ManagedObject<T> extends Serializable {
+
```
This part could be made a bit more null safe and shorter:
```dart
  late final Map<String?, Prop?> properties = entity.properties;

  late final Map<String?, Prop?> responseKeyProperties = {
    for (final key in properties.keys)
      if (key != null) mapKeyName(key): properties[key]
  };

  late final bool modelFieldIncludeIfNull = () {
    if (properties.isEmpty) {
      return true;
    }
    return properties.values.first?.responseModel?.fieldIncludeIfNull ?? true;
  }();

  String mapKeyName(String propertyName) {
    final property = properties[propertyName];
    return property?.responseKey?.name ?? property?.name ?? propertyName;
  }
```

Reductions commented on conduit/lib/src/db/managed/object.dart:
```diff
@@ -273,4 +313,16 @@ abstract class ManagedObject<T> extends Serializable {
   static bool _isPropertyPrivate(String propertyName) =>
       propertyName.startsWith("_");
+
+  bool _includeIfNull(ManagedPropertyDescription property) {
```
This functions could be:
```dart
  bool _includeIfNull(ManagedPropertyDescription property) {
    return property.responseKey?.includeIfNull ?? modelFieldIncludeIfNull;
  }
```

Reductions commented on conduit/lib/src/runtime/orm/entity_builder.dart:
```diff
@@ -39,10 +38,11 @@ class EntityBuilder {
   ManagedEntityRuntime? runtime;
 
-  String? name;
+  late String name;
```
```suggestion
  late final String name = _getName();
```

Reductions commented on conduit/lib/src/runtime/orm/entity_builder.dart:
```diff
@@ -1,23 +1,22 @@
-        metadata = firstMetadataOfType(getTableDefinitionForType(type)) {
+        metadata = firstMetadataOfType(getTableDefinitionForType(type),),
+        responseModel = firstMetadataOfType(getTableDefinitionForType(type),) {
     name = _getName();
```
```suggestion

```

Reductions reviewed (changes requested):
The annotation with the response keys is not a good design in my opinion.
It would be better for us to make a way to create a `MangedObeject` that is not tied to a table and use the DTO pattern.

luis901101 commented:
Hi, first of all thanks for the fast review and acknowledgment. I see `conduit` as a powerfull framework and since I discovered around 2 weeks ago I have been trying it and reading documentation to understand it the best I can and be able to use it in production projects.
Maybe I rushed a bit with this PR taking into account my lack of experience with `conduit`, but I like this framework a lot and I want to use it in real project as soon as possible and that encouraged me to do this PR…

@Reductions I refactored code with all of your suggestions, I have to confess I was unsure while deciding the more suitable names for this features mainly because a framework has a whole concept about naming features and I don´t know enough about `conduit` for making a good choice in this matter. Having said that, is more than welome any suggestions from you guys in fact feel free to do any changes you consider convenient.

About `legacyNaming` and `columnLegacyNaming` you were right it was not good naming at all for the properties in question. I used your suggestions with a minor variation, now is `useSnakeCaseName` and `useSnakeCaseColumnName`... this way feels more suitable... nevertheless I had to invert the imlementation of this properties due to the change of concept and I will have to change this PR description as well.

Finally about tests, I implemented more tests, but I was not able to run and pass all `conduit` tests... in fact I changed to master branch with no code changes at all and run tests but I get several tests failed, something about `wildfire` package not found... maybe it's something missconfigured in my local env, not sure. I checked the failed tests to see if it was somehting related to the new features and I didn't find anything related... anyway I think I will need some help in this tests matter.

luis901101 commented:
I just realized there are some internal tables and its columns that will not benefit from the `snake_case` naming implementation… I refer to the `AuthClient`, `AuthToken`, `PostgreSQLSchemaGenerator` tables *(and maybe others that I don’t know)*… 
Perhaps some static global flag can be used for this cases… 🤔

Reductions commented:
You are right about the internal tables. The thing you suggest is a good approach. Can you try to make a first draft of it.

luis901101 commented:
> @luis901101
> You are right about the internal tables. The thing you suggest is a good approach. Can you try to make the first draft of it?

Ok I will study the code involved and make a first approach implementation.

luis901101 commented:
@Reductions please check this notes...

For `snake_case` naming feature to benefit internal tables some important changes has to be made and notes to be taken into account:

1. `_ManagedAuthToken` and `_ManagedAuthClient` uses the `@Table(name: '_authtoken')` and `@Table(name: '_authclient')` annotations respectively, because the table names are set full lower case, there is no word separation distinction for snake_case to apply. So annotations needs to be updated to `@Table(name: '_AuthToken')` and `@Table(name: '_AuthClient')` respectively.

2. Because of the changes explained above is probably that `conduit db generate` fails to generate migration files on existing projects, with error: `Cannot change type of '_authtoken.client'` due to the different SchemaTable names used in `00000001_initial.migration.dart`. To solve this just update the initial migration file replacing every `_authclient` with `_AuthClient` and `_authtoken` with `_AuthToken`. Maybe `SchemaException` message could be updated to show this tip for this specific case.

3. Property `redirectURI` in `_ManagedAuthClient` should be `redirectUri` to ensure proper snake_case name `redirect_uri` otherwise would be `redirect_u_r_i`

4. When using `User` model as snake_case the table name would be `user`, creating a user table in postgres is not recommended because `user` is a keyword in postgres, in fact the migration upgrade fails because of this `CREATE TABLE user (id BIGSERIAL PRIMARY KEY,username TEXT NOT NULL UNIQUE,hashed_password TEXT NOT NULL,salt TEXT NOT NULL)`, that query throws `ERROR:  syntax error at or near "user"`. Of course developers could use a different name for the user model like `UserX` but happens that all conduit documentation and templates uses `User` for this model so the framework solution here should be to update `PostgreSQLSchemaGenerator` to use double quote for table names in every query e.g: `CREATE TABLE "user" (......)`, this way postgres will not complain about keywords. 

5. The lack of double quotes in the internal framework queries causes conflicts with Postgres keywrods as stated above and also causes table and column names to be full lowercase being different than the column or table name from Model. Updating all framework queries to use double quotes `"..."` for table and column names would be great to avoid name conflicts between model definitions and Postgres keywords **¡¡¡BUT!!!** doing this is not easy at all, there are lot of places where queries are used in the framework and more important any current project using conduit doing queries without double quotes will break due to the difference in the names.

luis901101 commented:
### Summarizing
After this notes and doing lot of tests and code review I think `snake_case` naming feature must be a breaking change *(as @Reductions mentioned here [Issue 67](https://github.com/conduit-dart/conduit/issues/67))*. Keeping `conduit` backward compatibile and also allowing devs to opt in to use `snake_case` naming for all tables and column names on conduit is really hard to get; not to mention the double quote refactoring.

j4qfrost commented:
Thanks a bunch you two. I can't believe it's been a year :/ closing in favor of this https://github.com/conduit-dart/conduit/pull/153
